### PR TITLE
add testing framework

### PR DESCRIPTION
**Pull request by yuunlimm**
### Description
Document to help understanding the components: https://www.notion.so/aptoslabs/SDK-Integration-Understanding-components-and-Writing-a-Test-Case-11a8b846eb72802aa023f40a3d0f8013?pvs=4

### Test Plan
https://github.com/aptos-labs/aptos-indexer-processors/pulls

### CODE REVIEW

File: aptos-indexer-processors-sdk/testing-framework/src/sdk_test_context.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+use aptos_protos::{indexer::v1::TransactionsResponse, transaction::v1::Transaction};
+use diesel::{pg::PgConnection, Connection};
+use testcontainers::{
+    core::{IntoContainerPort, WaitFor},
+    runners::AsyncRunner,
+    ContainerAsync, GenericImage, ImageExt,
+};
+use url::Url;
+
+pub struct SdkTestContext {
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/sdk_test_context.rs:16]: How difficult would it be to move the DB-specific setup to a trait? With SDK we've removed all references to diesel/Postgres and instead provide traits to let the processors define the DB dependency and queries. Wonder if we can abstract those parts in the testing framework as well so it could potentially work for any DB?

Comment (rtso) [reply aptos-indexer-processors-sdk/testing-framework/src/sdk_test_context.rs:16]: I _think_ it's doable, if we move the container setup and get_db_url into a trait. And change the `run` function to return the db url instead of the connection.

Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/sdk_test_context.rs:16]: that approach should not be overly difficult, so it would look like SdkTestContext to work with any test DB. something like this
trait 
```
pub trait TestDatabase: {
    fn establish_connection(&self) -> anyhow::Result<>
}
```

```
pub struct SdkTestContext<DB: TestDatabase> {
    pub transaction_batches: Vec<Transaction>,
    pub database: DB,
}

```
i am not sure what else needs to be abstracted away and might need some insights of defining database trait to be sufficiently flexible while it's still eays to use and working.

Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/sdk_test_context.rs:16]: assume they implemented that trait for Postgre, then they can write test like 

```
#[tokio::test]
async fn test_run() {
    let imported_txns = [IMPORTED_TESTNET_TXNS_1255836496_V2_FA_METADATA_];

    let db = PostgresTestDatabase {
        connection_string: "postgres://postgres:postgres@localhost/postgres".to_string(),
    };

    let test_context = SdkTestContext::new(&imported_txns, db)
        .await
        .unwrap();

    let events_processor_config = EventsProcessorConfig {
        channel_size: 10,
        per_table_chunk_sizes: AHashMap::new(),
    };
    let processor_config = ProcessorConfig::EventsProcessor(events_processor_config);
    let events_processor = test_context
        .init_events_processor(
            db.connection_string.clone(),
            processor_config,
            Some(1255836496),
            Some(1255836496),
        )
        .await;


    test_context
        .run(
            &events_processor,
            test_type,
            move |conn, _txn_version| {
                // Example validation logic
                conn.execute_transaction("1")?;
                Ok(())
            },
        )
        .await
        .unwrap();
}


```

Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/sdk_test_context.rs:16]: run function 
```
    pub async fn run<F>(
        &self,
        processor: &impl ProcessorTrait,
        test_type: TestType,
        verification_f: F,
    ) -> anyhow::Result<()>
    where
        F: Fn(Box<dyn DbConnection>, &str) -> anyhow::Result<()> + Send + Sync + 'static,
    {

        let db_conn = self.database.establish_connection()?;

        processor.run_processor().await?;

           test_type.run_verification(&db_conn, "1", &verification_f)?;
       

        Ok(())
    }
```

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        &self,
+        req: Request<GetTransactionsRequest>,
+    ) -> Result<Response<Self::GetTransactionsStream>, Status> {
+        let version = req.into_inner().starting_version.unwrap();
+        let transaction = self.transactions.iter().find(|t| {
+            t.transactions.iter().any(|tx| {
+                println!("Checking transaction version: {}", tx.version);
+                tx.version == version
+            })
+        });
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs:31]: Should we just always return `self.transactions`? It's kind of weird to only return one of the transactions. We can also make sure that `self.transactions` is sorted by version here.

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        // Start the Mock GRPC server
+        tokio::spawn(async move {
+            println!("Starting Mock GRPC server");
+            mock_grpc_server.run().await;
+        });
+    }
+
+    pub fn create_transaction_stream_config(&self,
+        starting_version: u64,
+        ending_version: u64,
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:110]: This could be cleaned up a bit imo, maybe this should accept a vec of transactions, and then dynamically sort them by version and get the starting/ending version from the vec?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+#[async_trait]
+pub trait TestDatabase: Send + Sync {
+    /// Set up the test container using user-defined code.
+    async fn setup<'a>(&'a mut self) -> anyhow::Result<()>;
+
+    /// Retrieve the database connection URL after setup.
+    fn get_db_url(&self) -> String;
+}
+
+pub struct PostgresTestDatabase {
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:137]: nit: move the Postgres related stuff into a new file.

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            mock_grpc_server.run().await;
+        });
+    }
+
+    pub fn create_transaction_stream_config(&self,
+        starting_version: u64,
+        ending_version: u64,
+    ) -> TransactionStreamConfig {
+        TransactionStreamConfig {
+            indexer_grpc_data_service_address: Url::parse("http://localhost:51254")
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:113]: nit: move this to a const

File: aptos-indexer-processors-sdk/transaction-stream/src/transaction_stream.rs
Comment (rtso) [inline aptos-indexer-processors-sdk/transaction-stream/src/transaction_stream.rs:1]: remove these changes?

File: rust/Cargo.lock
Comment (rtso) [inline rust/Cargo.lock:1]: This folder got renamed to aptos-indexer-processors-sdk, so this lock file is older/can be deleted

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+        // Pass the DB URL for user-defined validation
+        let db_url = self.database.get_db_url();
+        let db_values = verification_f(&db_url)?;
+
+        // Conditionally generate output files if the `generate_files` flag is true
+        if generate_files {
+            // Use the provided output path or fall back to default
+            let processor_name = processor.name();
+            let processor_name = "events_processor"; //  TODO: add a function to get name in runtime
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:68]: remove this?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+            let txn_version = txn_verion.to_string();
+            let output_dir = output_path.unwrap_or_else(|| {
+                format!(
+                    "expected_db_output_files/imported_testnet_txns/{}/{}_{}.json",
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:72]: Let's make the default output folder a const.

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+            database,
+        })
+    }
+
+    /// Run the processor and pass user-defined validation logic
+    pub async fn run<F>(
+        &self,
+        processor: &impl ProcessorTrait,
+        txn_verion: u64,
+        generate_files: bool, // flag to control file generation
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:42]: Instead of this being a bool that's passed in, could it just be inferred from the cargo test flags?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+            transaction_batches,
+            database,
+        })
+    }
+
+    /// Run the processor and pass user-defined validation logic
+    pub async fn run<F>(
+        &self,
+        processor: &impl ProcessorTrait,
+        txn_verion: u64,
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:41]: What if there's multiple versions in this test?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        generate_files: bool, // flag to control file generation
+        output_path: Option<String>, // Optional custom output path
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:43]: This should probably also better set using the CLI with cargo test flag

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+        // Run the processor
+        processor.run_processor().await?;
+
+        // Pass the DB URL for user-defined validation
+        let db_url = self.database.get_db_url();
+        let db_values = verification_f(&db_url)?;
+
+        // Conditionally generate output files if the `generate_files` flag is true
+        if generate_files {
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:65]: Will there also be a diff mode added later?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+    runners::AsyncRunner,
+    ContainerAsync, GenericImage, ImageExt,
+};
+use url::Url;
+use aptos_indexer_processor_sdk::aptos_indexer_transaction_stream::TransactionStreamConfig;
+use crate::mock_grpc::MockGrpcServer;
+
+pub struct SdkTestContext<D: TestDatabase> {
+    pub transaction_batches: Vec<Transaction>,
+    pub database: D, // Holds the database setup (user-defined)
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:18]: QQ about this struct? Why does it need to store the database? I wonder if it could be moved out of this struct and managed as its own struct? That would allow more flexible database setups.

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+    type GetTransactionsStream = ResponseStream;
+
+    async fn get_transactions(
+        &self,
+        req: Request<GetTransactionsRequest>,
+    ) -> Result<Response<Self::GetTransactionsStream>, Status> {
+        let version = req.into_inner().starting_version.unwrap();
+        let transaction = self.transactions.iter().find(|t| {
+            t.transactions.iter().any(|tx| {
+                println!("Checking transaction version: {}", tx.version);
```
Comment (dermanyang) [inline aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs:28]: Should we be using logs here instead of prints?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 9 above)
Comment (dermanyang) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:41]: +1 on using a vec here. O/w we'd have to call run multiple times in our test if we're using multiple versions

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+    pub fn new() -> Self {
+        PostgresTestDatabase {
+            postgres_container: None,
+            connection_string: String::new(),
+        }
+    }
+}
+
+#[async_trait]
+impl TestDatabase for PostgresTestDatabase {
```
Comment (dermanyang) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:152]: Since cargo runs tests in parallel by default, do we need to worry about dirty state in the testcontainers? Maybe we can pass in an additional optional `db_name` param to setup to have different DB names to ensure isolation?

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+impl MockGrpcServer {
+    pub async fn run(self) {
+        tonic::transport::Server::builder()
+            .add_service(
+                RawDataServer::new(self)
+                    .accept_compressed(tonic::codec::CompressionEncoding::Zstd)
+                    .send_compressed(tonic::codec::CompressionEncoding::Zstd),
+            )
+            .serve("127.0.0.1:51254".parse().unwrap())
```
Comment (dermanyang) [inline aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs:58]: Why this port?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+    }
+
+    /// Run the processor and pass user-defined validation logic
+    pub async fn run<F>(
+        &self,
+        processor: &impl ProcessorTrait,
+        txn_verion: u64,
+        generate_files: bool, // flag to control file generation
+        output_path: Option<String>, // Optional custom output path
+        verification_f: F,
```
Comment (dermanyang) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:44]: Could we have some explanation on how this verification part works? It was hard to follow in the sdk processor test 😬

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 14 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs]: Good point!!!

I think,  in current setup using `testcontainers`, each test case will run a separate database instance, because the `PostgresTestDatabase::setup` method starts a new Postgres container for each test. This should ensure that the tests are isolated at the container level, and each test will have its own instance of the database. but let me double-check i may be wrong lol

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 16 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:41]: good call out! will include this in a runbook/doc!

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 14 above)
Comment (dermanyang) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs]: Sounds good. I have trauma from this in the past lol

Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs]: haha me too lol. i added another test case, and fortunately not an issue!

Comment (dermanyang) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs]: awesome thanks!

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 9 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs]: as described above, added a todo. the way we mock result. doesn't support a txnStreamResponse with the gap.

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 10 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:63]: Added a cli args, but it still requires passing an arg here

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 11 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:79]: default mode will be a diff mode.

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs (hunk 13 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs]: cargo test doesn't print log as far as I know. but I may be wrong. will double check

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs (hunk 15 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs]: hmm no particular reason I copied this one from aptos-core lol

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 9 above)
Comment (dermanyang) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs]: Spoke offline; multiple versions will be supported in the test in the future. Maybe we can add a TODO here indicating that?

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs (hunk 15 above)
Comment (dermanyang) [reply aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs]: does the user need to specify this port to connect? maybe we can define it as a constant if so. Otherwise nw.

**Comment by dermanyang** [comment]
Looks good to me! Thanks for addressing all the comments. Leaving approval to @rtso

**Comment by yuunlimm** [comment]
I have more changes to support generating a scripted file with different naming convention. I will follow-up with the separate pr.

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs (hunk 15 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs]: no they don't have to, and the latest change should be using the constant

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    pub chain_id: u64,
+}
+
+type ResponseStream = Pin<Box<dyn Stream<Item = Result<TransactionsResponse, Status>> + Send>>;
+
+#[tonic::async_trait]
+impl RawData for MockGrpcServer {
+    type GetTransactionsStream = ResponseStream;
+
+    async fn get_transactions(
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs:25]: It seems like this only returns 1 transaction at a time? I think it could be helpful for the txn stream to emit more than 1 transaction, like if you want to test a sequence of transactions. Or if a processor processes txn's in parallel, txn's could get out of order, and we want to test that the DB output stays the same.

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 18]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    /// Helper function to set up and run the mock GRPC server.
+    async fn setup_mock_grpc(
+        &self,
+        transactions_response: Vec<TransactionsResponse>,
+        chain_id: u64,
+    ) {
+        let mock_grpc_server = MockGrpcServer {
+            transactions_response: Mutex::new(transactions_response),
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:142]: Why does this need a Mutex around the txn's?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 8 above)
Comment (rtso) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:70]: Could we parse the CLI args inside this function so the tests don't have to do it?

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+    // Return HashMap for multi-table results
+    where
+        F: FnOnce(&str) -> anyhow::Result<HashMap<String, serde_json::Value>>
+            + Send
+            + Sync
+            + 'static, // Modified for multi-table verification
+    {
+        let retry_strategy = ExponentialBackoff::from_millis(100).map(jitter).take(5); // Retry up to 5 times
+
+        let timeout_duration = TokioDuration::from_secs(10); // e.g., 5 seconds timeout for each retry
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:83]: How often does this fail? 😥

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+                eprintln!("[ERROR] Processor failed after retries: {:?}", e);
+                return Err(anyhow::anyhow!(
+                    "Failed to run processor after multiple retries: {}",
+                    e
+                ));
+            },
+        }
+
+        // Small delay to ensure all data is processed before verification
+        tokio::time::sleep(Duration::from_millis(250)).await;
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:108]: nit: put this in a const

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+        context.setup_mock_grpc(transactions_response, 1).await;
+
+        Ok(context)
+    }
+
+    /// Run the processor and pass user-defined validation logic
+    pub async fn run<F>(
+        &mut self,
+        processor: &impl ProcessorTrait,
+        db_url: &str,
```
Comment (rtso) [inline aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:68]: It looks like this `db_url` is only used to be passed into `verification_f`. My question is why does this `run` function need the `db_url` variable? The test could refer to the `db_url` inside of the function instead of having it passed down from another function.

**Review by rtso — COMMENTED** [review]
Testing framework is looking good! Left some comments

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 19 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:80]: after adding dynamic porting, it hasn't failed yet. but will keep it here just in case

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 21 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:65]: I left a comment about this, I agreed that we wouldn't need to pass in db_url here. as verification function can directly access db_url

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 18 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs]: this is. no longer needed as dynamic porting feature is added to mock grpc

File: aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs (hunk 17 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs:26]: I have a follow-up PR to handle this: https://github.com/aptos-labs/aptos-indexer-processor-sdk/pull/70

Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/mock_grpc.rs:26]: and out of order txns, I can also follow-up to permute

File: aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs (hunk 19 above)
Comment (yuunlimm) [reply aptos-indexer-processors-sdk/testing-framework/src/new_test_context.rs:82]: https://github.com/aptos-labs/aptos-indexer-processor-sdk/pull/70 this should handle.

**Review by dermanyang — APPROVED** [review]

**Review by rtso — APPROVED** [review]
